File: apps/web/src/engine/world/sync/tokenGrid.ts
/**
 * tokenGrid.ts — how many squares each token fills, as its game system
 * declares sizes.
 *
 * Spec 046 US4, FR-031 (research R10, contract §5). The server resolves each
 * token's size — a linked token's from its actor, a copy's from its NPC —
 * against the system's `combat.sizes`, and this hands every answer to the
 * engine through `set_token_grid`. That command has existed since spec 001,
 * and everything that should follow a creature's size already reads what it
 * sets: snapping, hit-testing, keyboard moves, nameplates and bars. Until now
 * nothing in the product sent it, so a Large ogre was a one-square token drawn
 * twice as big.
 *
 * Shaped like `tokenVision.ts`, for the same reasons: read rather than pushed,
 * because a size is on a *sheet* and a sheet edit is not a scene event; and
 * every token told, because a token left out would keep a size it lost.
 */
import { postGraphQL } from "@/api/graphqlClient";
import type { WorldStore } from "../store";
import { ACTOR_SHEET_CHANGED_EVENT_CODE } from "./tokenVision";

export type TokenGrid = {
  tokenId: string;
  /** Squares a side. */
  footprint: number;
};

type TokenGridQuery = { tokenGrid: TokenGrid[] };

/** The footprint of every token in this scene that fills other than one square. */
export async function getTokenGrid(sceneId: string): Promise<TokenGrid[]> {
  const answer = await postGraphQL<TokenGridQuery>(
    `
      query TokenGrid($sceneId: UUID!) {
        tokenGrid(sceneId: $sceneId) {
          tokenId
          footprint
        }
      }
    `,
    { sceneId },
  );
  return answer.tokenGrid;
}

/**
 * Read the scene's footprints and tell the engine.
 *
 * Tokens the server omits fill one square, and are told so explicitly —
 * `footprint: 1`. A creature whose size goes from Large to Medium has to
 * shrink on every board, and the engine would never hear that it had if the
 * token were simply left out.
 */
export async function loadTokenGridIntoEngine(
  worldStore: WorldStore,
  sceneId: string,
  tokenIds: Iterable<string>,
): Promise<void> {
  let sized: TokenGrid[];
  try {
    sized = await getTokenGrid(sceneId);
  } catch (error) {
    // A system that declares no sizes answers with an empty list, not an
    // error. An error is a real failure, and the honest response is to leave
    // every token as it is rather than to shrink them all to one square.
    console.error("Failed to read token sizes:", error);
    return;
  }

  const byToken = new Map(sized.map((grid) => [grid.tokenId, grid.footprint]));
  for (const tokenId of tokenIds) {
    worldStore.dispatch(
      {
        type: "set_token_grid",
        tokenId,
        footprint: byToken.get(tokenId) ?? 1,
      },
      "sync",
    );
  }
}

type WorldEventLike = { event_code?: number; eventCode?: number };

/**
 * Re-read the scene's footprints when any sheet changes (event 26).
 *
 * Event 14 (a token changed) is handled where the tokens are re-read
 * (`tokens.ts`), after them, so a new token is in the engine before it is
 * sized. This joins `WorldPage`'s one fan-out for the sheet event, as
 * `applyTokenVisionWorldEvent` does.
 */
export async function applyTokenGridWorldEvent(
  worldStore: WorldStore,
  sceneId: string,
  event: WorldEventLike,
  tokenIdsNow: () => Iterable<string>,
): Promise<void> {
  const code = event.event_code ?? event.eventCode;
  if (code !== ACTOR_SHEET_CHANGED_EVENT_CODE) {
    return;
  }
  await loadTokenGridIntoEngine(worldStore, sceneId, tokenIdsNow());
}

File: src/server/src/graphql/queries/token_grid.rs
//! `tokenGrid(sceneId)` — how many squares each token in a scene fills (spec
//! 046 US4, contract §1, research R10).
//!
//! A sibling query keyed by token id, like `tokenVision`, for the same reason:
//! the answer comes from a system's manifest and every actor's sheet, loaded
//! once for the scene. The web hands each answer to the engine's
//! `set_token_grid`, which everything that should follow a creature's size —
//! snapping, hit-testing, keyboard moves, nameplates, bars — already reads.

use async_graphql::{Context, Error, Object, Result as GraphQLResult, SimpleObject};
use diesel::prelude::*;
use uuid::Uuid;

use crate::combat::size::{DEFAULT_FOOTPRINT, footprints_in_scene};
use crate::graphql::{app_state, authenticated_user};

/// One token's footprint, in squares a side.
#[derive(SimpleObject, Debug, Clone)]
#[graphql(name = "TokenGrid")]
pub struct GraphQLTokenGrid {
    pub token_id: Uuid,
    pub footprint: f64,
}

#[derive(Default)]
pub struct TokenGridQuery;

#[Object]
impl TokenGridQuery {
    /// Every token in the scene that fills something other than one square:
    /// a linked token by its actor's size, a copy by its NPC's, as the world's
    /// game system declares sizes (`combat.sizes`).
    ///
    /// A token of one square is **omitted**. The client tells the engine `1`
    /// for every token it does not find here, so a creature that shrinks back
    /// to Medium shrinks on every board.
    async fn token_grid(
        &self,
        ctx: &Context<'_>,
        scene_id: Uuid,
    ) -> GraphQLResult<Vec<GraphQLTokenGrid>> {
        let user = authenticated_user(ctx)?;
        let (user_id, is_admin) = (user.user_id, user.is_admin);
        let state = app_state(ctx)?;
        let systems_dir = state.directories.systems_dir.clone();
        let mut conn = state
            .db_pool
            .get()
            .map_err(|_| Error::new("Failed to get DB connection"))?;

        tokio::task::spawn_blocking(move || {
            use crate::schema::scenes;

            let world_id: Uuid = scenes::table
                .filter(scenes::scene_id.eq(scene_id))
                .select(scenes::world_id)
                .first(&mut conn)
                .map_err(|_| Error::new("Scene not found"))?;
            let actor = crate::auth::world_membership::actor_in_world(
                &mut conn, user_id, is_admin, world_id,
            );
            if actor.role.is_none() && !actor.is_site_admin {
                return Err(Error::new("Not a member of this world"));
            }

            let footprints = footprints_in_scene(&mut conn, &systems_dir, scene_id)
                .map_err(|e| Error::new(format!("Failed to read token sizes: {e}")))?;
            let mut out: Vec<GraphQLTokenGrid> = footprints
                .into_iter()
                .filter(|(_, footprint)| *footprint != DEFAULT_FOOTPRINT)
                .map(|(token_id, footprint)| GraphQLTokenGrid {
                    token_id,
                    footprint: footprint as f64,
                })
                .collect();
            out.sort_by_key(|grid| grid.token_id);
            Ok(out)
        })
        .await
        .map_err(|e| Error::new(format!("Task failed: {e}")))?
    }
}
